app/scripts/modules/core/pipeline/config/stages/unmatchedStageTypeStage/unmatchedStageTypeStage.controller.ts

```diff
@@ -0,0 +1,93 @@
+import {module, IComponentController, IScope} from 'angular';
+import {JSON_UTILITY_SERVICE, JsonUtilityService} from 'core/utils/json/json.utility.service';
+import {IStage} from 'core/domain/IStage';
+import {IPipeline} from 'core/domain/IPipeline';
+
+interface IUnmatchedStageTypeStageScope extends IScope {
+  pipeline: IPipeline;
+  stage: IStage;
+}
+
+export class UnmatchedStageTypeStageCtrl implements IComponentController {
+  public stageJSON: string;
+  public errors: string[];
+  public errorMessage: string;
+  public textareaRows: number;
+
+  static get $inject() { return ['$scope', 'jsonUtilityService']; }
+
+  constructor(public $scope: IUnmatchedStageTypeStageScope, private jsonUtilityService: JsonUtilityService) { }
+
+  public $onInit(): void {
+    this.$scope.$watch('stage', () => this.setStageJSON(), true);
+    this.setStageJSON();
+    this.textareaRows = this.stageJSON.split('\n').length;
+  }
+
+  public updateStageJSON(): void {
+    let parsedStage: IStage,
+      errors: string[] = [];
+    try {
+      parsedStage = JSON.parse(this.stageJSON);
+    } catch (e) {
+      errors.push(e.message);
+    }
+    if (parsedStage) {
+      this.validateStage(parsedStage, errors);
+      this.validatePipeline(parsedStage, errors);
+    }
+
+    if (!errors.length) {
+      Object.keys(this.$scope.stage).forEach(key => delete this.$scope.stage[key]);
+      Object.assign(this.$scope.stage, parsedStage);
+      this.setStageJSON();
+    }
+
+    this.errors = errors;
+    this.renderErrors();
+  }
+
+  private setStageJSON(): void {
+    this.stageJSON = this.jsonUtilityService.makeSortedStringFromObject(this.$scope.stage);
```

Comment:
If you remove `refId` from the `stageJSON` here, would that obviate the need for all the validation of the `refId` field in the controller? You'd need to preserve it in `updateStageJSON` when deleting all the existing properties, but otherwise might make the logic simpler.